File: demo/src/main/java/design/pattern/lg/AbstractFactoryDemo.java
package design.pattern.lg;

class Chair {
}

class Sofa {
}

class Table {
}

class ChinaChair extends Chair {
}

class ChinaSofa extends Sofa {
}

class ChinaTable extends Table {
}

class USAChair extends Chair {
}

class USASofa extends Sofa {
}

class USATable extends Table {
}

class Client {
    private Chair myChair;
    private Sofa mySofa;
    private Table myTable;


    //通过抽象工厂来生产家具
    public Client(AbstractFactory af) {
        myChair = af.createChair();
        mySofa = af.createSofa();
        myTable = af.createTable();
    }
}

//抽象的家具工厂
abstract class AbstractFactory {
    abstract Chair createChair();

    abstract Sofa createSofa();

    abstract Table createTable();
}

//中国的家具工厂

class ChinaFactory extends AbstractFactory {

    @Override
    Chair createChair() {
        return new ChinaChair();
    }

    @Override
    Sofa createSofa() {
        return new ChinaSofa();
    }

    @Override
    Table createTable() {
        return new ChinaTable();
    }
}

//美国的家具工厂

class USAFactory extends AbstractFactory {

    @Override
    Chair createChair() {
        return new USAChair();
    }

    @Override
    Sofa createSofa() {
        return new USASofa();
    }

    @Override
    Table createTable() {
        return new USATable();
    }
}

File: demo/src/main/java/design/pattern/lg/AbstractFactoryDemo.kt
package design.pattern.lg

/**
 * 抽象工厂模式 demo
 */

internal open class Chair
internal open class Sofa
internal open class Table
internal class ChinaChair : Chair()
internal class ChinaSofa : Sofa()
internal class ChinaTable : Table()
internal class USAChair : Chair()
internal class USASofa : Sofa()
internal class USATable : Table()
internal class Client(af: AbstractFactory) {
    private val myChair: Chair
    private val mySofa: Sofa
    private val myTable: Table

    //通过抽象工厂来生产家具
    init {
        myChair = af.createChair()
        mySofa = af.createSofa()
        myTable = af.createTable()
    }
} //抽象的家具工厂

internal abstract class AbstractFactory {
    abstract fun createChair(): Chair
    abstract fun createSofa(): Sofa
    abstract fun createTable(): Table
} //中国的家具工厂

internal class ChinaFactory : AbstractFactory() {
    override fun createChair(): Chair {
        return ChinaChair()
    }

    override fun createSofa(): Sofa {
        return ChinaSofa()
    }

    override fun createTable(): Table {
        return ChinaTable()
    }
} //美国的家具工厂

internal class USAFactory : AbstractFactory() {
    override fun createChair(): Chair {
        return USAChair()
    }

    override fun createSofa(): Sofa {
        return USASofa()
    }

    override fun createTable(): Table {
        return USATable()
    }
}